openedx/core/lib/xblock_keys/bundle_def.py

```diff
@@ -0,0 +1,139 @@
+import base64
+from uuid import UUID
+
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import DefinitionKey
+
+
+class BundleDefinitionLocator(DefinitionKey):
+    """
+    Implementation of the DefinitionKey type, for XBlock content stored in
+    Blockstore bundles (.olx files).
+
+    Each BundleDefinitionLocator holds the following data
+        bundle UUID (e.g. fc4d4d8d-60b3-4aad-b9f1-4ea738505d13)
+        OLX file path (e.g. /unit-welcome/welcome.olx)
+        Block type (of some block within the OLX file, e.g. 'html')
+        Block definition ID (of some block within the OLX file, e.g. 'intro')
+    
+    Note that we do not store bundle slugs (rather UUIDs), because the slugs
+    may change. These definition keys are more for internal use than for
+    displaying to users.
+    """
+    CANONICAL_NAMESPACE = 'olx-v1'
+    KEY_FIELDS = ('bundle_uuid', 'olx_path', 'block_type', 'definition_id')
+    __slots__ = KEY_FIELDS
+    CHECKED_INIT = False
+
+    def __init__(self, bundle_uuid, olx_path, block_type, definition_id):
+        if not isinstance(bundle_uuid, UUID):
+            bundle_uuid = UUID(bundle_uuid)
+        if len(olx_path) < 2 or olx_path[0] != '/' or olx_path.find(':') != -1:
+            raise ValueError("Invalid OLX path: {}".format(olx_path))
+        super(BundleDefinitionLocator, self).__init__(
+            bundle_uuid=bundle_uuid,
+            olx_path=olx_path,
+            block_type=block_type,
+            definition_id=definition_id,
+        )
+
+    @classmethod
+    def _olx_path_shortened(cls, olx_path):
+        """
+        Convert an arbitrary OLX path to a potentially shorter format.
+        The result is shorter if the olx path is in the standard format
+        /block_type-block_id/block_id.olx
+
+        The result is a string like "block_type/block_id" or the original string.
+
+        Returns a tuple: short_string, type_part or None, id_part or None
+        """
+        if len(olx_path) < 2 or olx_path[0] != '/':
+            raise ValueError("An OLX path must start with / and contain a filename")
+        if olx_path.endswith('.olx'):
+            id_part = olx_path[:-4].rpartition('/')[2]  # from "/foo/bar.olx", get "bar"
+            type_part = olx_path[1:].partition('-' + id_part + '/')[0] # from "/foo-bar/bar.olx", get "foo"
+            short_path = type_part + '/' + id_part
+            if cls._olx_path_expanded(short_path)[0] == olx_path:
+                return (short_path, type_part, id_part)
```

Comment:
1. Isn't this standard format a bit too restrictive? I understand that in the course exports, blocks are put into directories names after their type, but do we need to replicate that? An average bundle is going to have far fewer files to require organization of files into sub-directories. Can we do just just `/<block_uuid>.olx`? This still allows doing arbitrary organization if required: `/<some/arbitrary/path/<block_uuid>.olx`.
2. Is it important to have the `block_type` in the path? Like for author friendliness? If yes, maybe we can move the type to the suffix: `/<block_uuid>.<block_type>.olx`?

In both cases we will just have to strip off everything after the first `.` and there will be an obvious correspondence between the path and the locator?